### PR TITLE
batch size tests

### PR DESCRIPTION
**Pull request by liqunfu**
**Description**: test memory usage w.r.t. training configurations

**Motivation and Context**
improve test coverage

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    cmds = [
+        sys.executable,
+        'orttraining_run_bert_pretrain.py',
+        "ORTBertPretrainTest.test_pretrain_throughput",
+        "--sequence_length", str(config.sequence_length),
+        "--max_batch_size", str(config.max_batch_size),
+        "--max_predictions_per_seq", str(config.max_predictions_per_seq)]
+    if config.enable_mixed_precision:
+        cmds = [*cmds, "--enable_mixed_precision"]
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py:44]: nit: how about `cmds += ["..."]`

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 from dataclasses import dataclass, field
 from typing import Optional, Any, Dict
 import json
+import collections
```
Comment (edgchen1) [inline orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:17]: used?

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -1163,28 +1163,27 @@ def run_training_python_frontend_e2e_tests(cwd):
     # frontend tests are to be added here:
     log.info("Running python frontend e2e tests.")
 
+    run_subprocess(
```
Comment (edgchen1) [inline tools/ci_build/build.py:1166]: i think build.py has a lot of stuff in it already. can this script be invoked from a build definition instead?
or maybe we should rename build.py :)

Comment (liqunfu) [reply tools/ci_build/build.py:1163]: orttraining-linux-gpu-e2e-test-ci-pipeline.yml is a good example for doing e2e tests in a build definition. I see many benefits of it. I will do as you suggested next when merging cpp e2e and python e2e pipelines into one pipeline.

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py (hunk 2 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:18]: removed, thanks!

File: orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py (hunk 1 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py:44]: used list.append. += is not working as it treat "..." as a list of chars.

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py [hunk 4]
```diff
@@ -524,41 +585,41 @@ def setUp(self):
...
-            learning_rate=self.learning_rate,
-            max_seq_length=self.max_seq_length,
-            max_predictions_per_seq=self.max_predictions_per_seq,
-            train_batch_size=self.train_batch_size,
-            gradient_accumulation_steps=self.gradient_accumulation_steps,
-            input_dir=self.input_dir,
-            fp16=self.fp16,
-            allreduce_post_accumulation=self.allreduce_post_accumulation)
+            max_steps=10,
+            learning_rate=5e-4,
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:610]: should these be hard-coded?

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py [hunk 5]
```diff
@@ -524,41 +585,41 @@ def setUp(self):
...
+            max_predictions_per_seq=process_args.max_predictions_per_seq,
+            train_batch_size=process_args.max_batch_size,
+            gradient_accumulation_steps=1,
+            fp16=process_args.enable_mixed_precision,
+            gelu_recompute=process_args.gelu_recompute,
+            attn_dropout_recompute=process_args.attn_dropout_recompute,
+            transformer_layer_recompute=process_args.transformer_layer_recompute,
+            allreduce_post_accumulation=True,
+            # TODO: remove
+            force_num_hidden_layers=2,
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:621]: Should these be hard-coded?

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py [hunk 6]
```diff
@@ -524,41 +585,41 @@ def setUp(self):
...
+            input_dir = '/bert_data/hdf5_lower_case_1_seq_len_512_max_pred_80_masked_lm_prob_0.15_random_seed_12345_dupe_factor_5/books_wiki_en_corpus/train'
 
-        logger.info("self.gradient_accumulation_steps = %d", self.gradient_accumulation_steps)
+        print("process_args.enable_mixed_precision: ", process_args.enable_mixed_precision)
+        print("process_args.sequence_length: ", process_args.sequence_length)
+        print("process_args.max_batch_size: ", process_args.max_batch_size)
+        print("process_args.max_predictions_per_seq: ", process_args.max_predictions_per_seq)
+        print("process_args.gelu_recompute: ", process_args.gelu_recompute)
+        print("process_args.attn_dropout_recompute: ", process_args.attn_dropout_recompute)
+        print("process_args.transformer_layer_recompute: ", process_args.transformer_layer_recompute)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:600]: Are these `print` just for debugging or meant to be in the final PR?

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py [hunk 7]
```diff
@@ -673,19 +738,7 @@ def test_pretrain_zero(self):
...
-        elif len(sys.argv) >= 2 and sys.argv[1] == 'ORTBertPretrainTest.test_pretrain_convergence':
-            logger.info("running ORTBertPretrainTest.test_pretrain_convergence()...")
-            test.max_steps = 200
-            test.force_num_hidden_layers = 8
-            final_loss = test.test_pretrain_convergence()
-            logger.info("ORTBertPretrainTest.test_pretrain_convergence() final loss = %f", final_loss)
-            test.assertLess(final_loss, 8.5)
-            logger.info("ORTBertPretrainTest.test_pretrain_convergence() passed")
-        elif len(sys.argv) >= 2 and sys.argv[1] == 'ORTBertPretrainTest.test_pretrain_zero':
+        if len(sys.argv) >= 2 and sys.argv[1] == 'ORTBertPretrainTest.test_pretrain_zero':
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:741]: Consider using `argparse` as opposed to manipulate `argv` directly

File: orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    Config(True, 512, 9, 80, False, True, False),
+    Config(True, 512, 15, 80, False, False, True),
+]
+
+def run_with_config(config):
+    print("##### testing name - {}-{} #####".format("fp16" if config.enable_mixed_precision else "fp32",
+                                                        config.sequence_length))
+    print("gelu_recompute: ", config.gelu_recompute)
+    print("attn_dropout_recompute: ", config.attn_dropout_recompute)
+    print("transformer_layer_recompute: ", config.transformer_layer_recompute)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py:34]: Are these intended or debugging?

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py (hunk 4 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:610]: As a nature of unit test (I did not fully agree though), it is not recommended to pass arguments as test parameters. Thus it need to be hard coded.

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py (hunk 5 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:621]: same as last comment

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py (hunk 7 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:746]: unit test is against passing arguments so we need sys.argv to work around it. As you can see I still do not have a good solutions for it - I am not sure if a solution is needed for this artificial restriction of not allowing arguments to an unit test.

File: orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py (hunk 8 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_run_frontend_batch_size_test.py:34]: It is helpful to see what is being tested. Yes it is intended.

File: orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py (hunk 6 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_run_bert_pretrain.py:600]: Yes these prints is to reconfirm that the configuration to be used for running the trainer.

**Review by thiagocrepaldi — APPROVED** [review]

**Pull request merged**